### PR TITLE
+ Redirect URL Storage

### PR DESCRIPTION
**Pull request by maqroll**
I hereby agree to the terms of the CLA available at: https://yandex.ru/legal/cla/?lang=en

Category (leave one):
- Improvement

Short description (up to few sentences):

Support (optional) redirects on URL storage.

Detailed description (optional):

Set up max number of GET redirects on URL storage using max_http_get_redirects.
max_http_get_redirects=0 (default) means no redirects allowed.

### CODE REVIEW

File: dbms/src/IO/ReadWriteBufferFromHTTP.cpp [hunk 1]
```diff
@@ -1 +1,40 @@
...
+		UInt64 redirects = 0;
+		
+		do 
+		{
+			try 
+			{
+				return std::make_unique<DB::ReadWriteBufferFromHTTP>(actual_uri, method, callback, timeouts);
+			} 
+			catch (Poco::URIRedirection & exc) {
+				redirects++;
```
Comment (alexey-milovidov) [inline dbms/src/IO/ReadWriteBufferFromHTTP.cpp]: try/catch is not a good way to implement application logic.

**Comment by alexey-milovidov** [comment]
Security considerations should be noted in setting description.

File: dbms/src/Common/ErrorCodes.cpp [hunk 2]
```diff
@@ -452,6 +452,10 @@ namespace ErrorCodes
     extern const int INVALID_WITH_FILL_EXPRESSION = 475;
     extern const int WITH_TIES_WITHOUT_ORDER_BY = 476;
     extern const int INVALID_USAGE_OF_INPUT = 477;
+    extern const int TOO_MANY_REDIRECTS = 478;
+ 
+    extern const int KEEPER_EXCEPTION = 999;
```
Comment (maqroll) [inline dbms/src/Common/ErrorCodes.cpp]:
```suggestion
```

File: dbms/src/IO/ReadWriteBufferFromHTTP.cpp [hunk 3]
```diff
@@ -1 +1,40 @@
...
+{
+
+namespace ErrorCodes
+{
+	extern const int TOO_MANY_REDIRECTS; 
+}
+
+
+std::unique_ptr<DB::ReadWriteBufferFromHTTP> makeReadWriteBufferFromHTTP(const Poco::URI & uri,
+	const std::string & method,
```
Comment (maqroll) [inline dbms/src/IO/ReadWriteBufferFromHTTP.cpp]:
```suggestion
    const std::string & method,
```

File: dbms/src/IO/ReadWriteBufferFromHTTP.cpp [hunk 4]
```diff
@@ -1 +1,40 @@
...
+	std::function<void(std::ostream &)> callback,
+	const DB::ConnectionTimeouts & timeouts,
```
Comment (maqroll) [inline dbms/src/IO/ReadWriteBufferFromHTTP.cpp]:
```suggestion
    const DB::ConnectionTimeouts & timeouts,
```

File: dbms/src/IO/ReadWriteBufferFromHTTP.cpp [hunk 5]
```diff
@@ -1 +1,40 @@
...
+	const DB::SettingUInt64 max_redirects)
+	{
```
Comment (maqroll) [inline dbms/src/IO/ReadWriteBufferFromHTTP.cpp]:
```suggestion
    {
```

File: dbms/src/IO/ReadWriteBufferFromHTTP.cpp [hunk 6]
```diff
@@ -1 +1,40 @@
...
+		auto actual_uri =uri;
```
Comment (maqroll) [inline dbms/src/IO/ReadWriteBufferFromHTTP.cpp]:
```suggestion
        auto actual_uri =uri;
```

File: dbms/src/IO/ReadWriteBufferFromHTTP.cpp (hunk 1 above)
Comment (maqroll) [reply dbms/src/IO/ReadWriteBufferFromHTTP.cpp]: I agree of course. I considered that the other approach I took into account (include the number of redirects in the constructor, change receiveResponse and implement the redirection loop in the ReadWriteBufferFromHTTPBase constructor) had more risk (many more changes) and that the functionality implemented was not worth the change having a much simpler solution (without risk) although not as elegant.

**Comment by nikitamikhaylov** [comment]
If you are interested in it, that's ok for me :)

**Comment by nikitamikhaylov** [comment]
@maqroll  How things going? Do you need my help? Maybe it is better to reassign this task to me?

**Comment by maqroll** [comment]
> @maqroll How things going? Do you need my help? Maybe it is better to reassign this task to me?

@nikitamikhaylov If it's ok I'll report the status at the end of the day (in Spain) and you'll decide.

**Comment by maqroll** [comment]
@nikitamikhaylov  I got here (no experience in C ++). Maybe you can use it, it depends entirely on you. Thanks.

**Comment by nikitamikhaylov** [comment]
Ok.

**Comment by maqroll** [comment]
@nikitamikhaylov Last change compiles and works on basic tests.

**Comment by nikitamikhaylov** [comment]
Ok, I see. But we need all test to be green. I will do it in some time.

**Comment by maqroll** [comment]
Working in the integration test.

**Comment by maqroll** [comment]
@nikitamikhaylov @alexey-milovidov get rid of try/catch & added integration test

File: dbms/src/IO/ReadWriteBufferFromHTTP.h [hunk 7]
```diff
@@ -26,69 +27,156 @@ namespace DB
...
             try
             {
-                auto & stream_out = session->sendRequest(request);
+                auto & stream_out = sess->sendRequest(request);
 
                 if (out_stream_callback)
                     out_stream_callback(stream_out);
 
-                istr = receiveResponse(*session, request, response);
+                istr = receiveResponse(*sess, request, response, true);
```
Comment (alexey-milovidov) [inline dbms/src/IO/ReadWriteBufferFromHTTP.h]: That's suspicious, need double check it...

**Review by alexey-milovidov — APPROVED** [review]

File: dbms/src/IO/ReadWriteBufferFromHTTP.h (hunk 7 above)
Comment (maqroll) [reply dbms/src/IO/ReadWriteBufferFromHTTP.h]: receiveResponse before the change only accepted 200 code responses. Last parameter accepts redirects codes (307,...) as ok. That loop accepts redirects and checks max_redirects limit.

**Pull request merged**